midi_trainable_tokenizers/bpe_tokenizer.py

```diff
@@ -0,0 +1,106 @@
+import json
+from concurrent.futures import ThreadPoolExecutor
+
+import tokenizers
+import pandas as pd
+from datasets import Dataset
+from tokenizers import Regex, Tokenizer, models, trainers, pre_tokenizers
+
+from midi_tokenizers.midi_tokenizer import MidiTokenizer
+from object_generators.base_tokenizer_generator import BaseTokenizerGenerator
+from midi_trainable_tokenizers.trainable_tokenizer import MidiTrainableTokenizer
+
+
+class BpeMidiTokenizer(MidiTrainableTokenizer):
+    def __init__(self, base_tokenizer: MidiTokenizer, bpe_tokenizer: Tokenizer = None, max_vocab_size: int = None):
+        super().__init__()
+        self.base_tokenizer = base_tokenizer
+        self.tokenizer = bpe_tokenizer
+        self.name = "BpeTokenizer"
+        self.max_vocab_size = max_vocab_size
+
+        if self.max_vocab_size is None:
+            self.max_vocab_size = 30000  # default BpeTrainer vocab_size
+
+        if self.tokenizer is None:
+            # Initialize empty tokenizer and a trainer
+            self.tokenizer = Tokenizer(model=models.BPE())
+            self.tokenizer.pre_tokenizer = self.prepare_text_pre_tokenizer()
+            self.tokenizer.model = models.BPE()
+            self.trainer = trainers.BpeTrainer(
+                vocab_size=self.max_vocab_size,
+                max_token_length=512,
+                special_tokens=["<CLS>"],
+            )
```

Comment:
How can we guarantee that the BPE will not try to split "NOTE" into "NO" and "TE"? I supsect that's one of the pre-tokenizers functionalities, but I can't tell just by reading this code

Reply:
It will try to split it if it if wants to minimize the vocabulary (to eg. NOTE and _OFF/_ON), but we use this algorithm to make our vocabulary bigger, so there will be no need to further slice the note tokens.

Reply:
I'm still struggling to understand what's the relation of pre-tokenizers and BPE here - can you please prepare a google doc explaining with samples every step of the tokenization process that this implements?

Reply:
:+1: